xclim/core/bootstrapping.py

```diff
@@ -0,0 +1,236 @@
+    3. Get the thershold (your percentile of interest) form this 30 years in-base period
+    4. Calculate the exceedance rate of the out base period (the single year)
+    5. Repeat 28 times the steps 2, 3, 4 but each time, the repeated year of step 2 must be another year of the in-base
+    6. To obtain the final time serie of exceedance rate for the out of base year, average the 29 estimates obtained
+
+    7. Repeat the whole process for each year of the base period
+
+    TODO: show how to bootstrap a custom indice
+
+    """
+    in_base_period = da.sel(time=config.in_base_slice)
+    in_base_exceedance_rates = _bootstrap_period(
+        in_base_period, config, exceedance_function
+    )
+    if config.out_of_base_slice is None:
+        return in_base_exceedance_rates
+    out_of_base_period = da.sel(time=config.out_of_base_slice)
+    in_base_threshold = _calculate_thresholds(in_base_period, config)
+    out_of_base_exceedance = _calculate_exceedances(
+        config, exceedance_function, out_of_base_period, in_base_threshold
+    )
+    return xr.concat([in_base_exceedance_rates, out_of_base_exceedance], dim="time")
+
+
+def _bootstrap_period(
+    ds_in_base_period: DataArray,
+    config: BootstrapConfig,
+    exceedance_function: ExceedanceFunction,
+) -> DataArray:
+    period_exceedance_rates = []
+    for year_ds in ds_in_base_period.groupby("time.year"):
+        period_exceedance_rates.append(
+            _bootstrap_year(ds_in_base_period, year_ds[0], config, exceedance_function)
+        )
+    out = xr.concat(period_exceedance_rates, dim="time")
+    # workaround to ensure unit is really "days"
+    out.attrs["units"] = "d"
+    return out
+
+
+def _bootstrap_year(
+    ds_in_base_period: DataArray,
+    out_base_year: int,
+    config: BootstrapConfig,
+    exceedance_function: ExceedanceFunction,
+) -> DataArray:
+    print("out base :", out_base_year)
+    in_base = _build_virtual_in_base_period(ds_in_base_period, out_base_year)
+    out_base = ds_in_base_period.sel(time=str(out_base_year))
+    exceedance_rates = []
+    for year_ds in in_base.groupby("time.year"):
+        print(year_ds[0])
+        replicated_year = in_base.sel(time=str(year_ds[0]))
+        # it is necessary to change the time of the replicated year
+        # in order to not skip it in percentile calculation
+        replicated_year["time"] = replicated_year.time + pd.Timedelta(
+            str(out_base_year - year_ds[0]) + "y"
+        )
+        completed_in_base = xr.concat([in_base, replicated_year], dim="time")
+        thresholds = _calculate_thresholds(completed_in_base, config)
+        exceedance_rate = _calculate_exceedances(
+            config, exceedance_function, out_base, thresholds
+        )
+        exceedance_rates.append(exceedance_rate)
+    if len(exceedance_rates) == 1:
+        return exceedance_rates[0]
+    return xr.concat(exceedance_rates, dim="time").groupby("time").mean()
```

Comment:
I'm not sure what is happening here. Why would you `groupby('time')`? I understand that `exceedance_rate` is an array at sampling freq `freq` _only_ for the `out_base_year`, no? So, there are no overlap of time elements and `groupby("time").mean()` shouldn't have any effect, no?

Reply:
So it has the same time coordinate repeated multiple times because it is computed over the `out_base_period` for each in_base years.  I would then suggest:

```suggestion
    return xr.concat(exceedance_rates, dim="in_base_year").mean('in_base_year')
```
Again because of my chronic avoidance of `groupby` ;).